### PR TITLE
Add TempleScan

### PR DESCRIPTION
**Pull request by MikeZeDev**
no more HeanCMS, custom site.

### CODE REVIEW

File: web/src/engine/websites/TempleScan.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public override async FetchMangas(provider: MangaPlugin): Promise<Manga[]> {
+        const data = await FetchJSON<APIManga[]>(new Request(new URL('allComics', this.apiUrl)));
+        return data.map(manga => new Manga(this, provider, manga.series_slug, manga.title.trim()));
+    }
+
+    public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
+        const { Season } = await FetchJSON<APIManga>(new Request(new URL(`comic/${manga.Identifier}`, this.apiUrl)));
+        const results = Season.reduce((accumulator: Chapter[], element) => {
```
Comment (ronny1982) [inline web/src/engine/websites/TempleScan.ts:55]: Based on the type definition `Season` may be `null`.
Either check for `null` or change the type definition.

File: web/src/engine/websites/TempleScan.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    chapter_data: {
+        images: string[]
+    }
+}
+
+@Common.ImageAjax(true)
+export default class extends DecoratableMangaScraper {
+
+    private readonly apiUrl = 'https://api.templescan.net/api/';
+    private readonly mangaRegexp = new RegExp(`^${this.URI.origin}/(comic/[^/]+)$`);
```
Comment (ronny1982) [inline web/src/engine/websites/TempleScan.ts:29]: May use `RegExpSafe` instead.

File: web/src/engine/websites/TempleScan.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        return icon;
+    }
+
+    public override ValidateMangaURL(url: string): boolean {
+        return this.mangaRegexp.test(url);
+    }
+
+    public override async FetchManga(provider: MangaPlugin, url: string): Promise<Manga> {
+        const { series_slug, title } = await FetchJSON<APIManga>(new Request(new URL(url.match(this.mangaRegexp)[1], this.apiUrl)));
+        return new Manga(this, provider, series_slug, title.trim());
```
Comment (ronny1982) [inline web/src/engine/websites/TempleScan.ts:45]: The `trim` is probably redundant as JSON is not expected to contain whitespaces in contrary to HTML.

File: web/src/engine/websites/TempleScan.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        const data = await FetchJSON<APIManga[]>(new Request(new URL('allComics', this.apiUrl)));
+        return data.map(manga => new Manga(this, provider, manga.series_slug, manga.title.trim()));
+    }
+
+    public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
+        const { Season } = await FetchJSON<APIManga>(new Request(new URL(`comic/${manga.Identifier}`, this.apiUrl)));
+        const results = Season.reduce((accumulator: Chapter[], element) => {
+            const chapters = element.Chapter.map(chapter => {
+                const title = chapter.chapter_title ? [chapter.chapter_name, chapter.chapter_title].join(' : ') : chapter.chapter_name;
+                return new Chapter(this, manga, chapter.chapter_slug, title.trim());
```
Comment (ronny1982) [inline web/src/engine/websites/TempleScan.ts:58]: The `trim` is probably redundant as JSON is not expected to contain whitespaces in contrary to HTML.

File: web/src/engine/websites/TempleScan.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public override async FetchManga(provider: MangaPlugin, url: string): Promise<Manga> {
+        const { series_slug, title } = await FetchJSON<APIManga>(new Request(new URL(url.match(this.mangaRegexp)[1], this.apiUrl)));
+        return new Manga(this, provider, series_slug, title.trim());
+    }
+
+    public override async FetchMangas(provider: MangaPlugin): Promise<Manga[]> {
+        const data = await FetchJSON<APIManga[]>(new Request(new URL('allComics', this.apiUrl)));
+        return data.map(manga => new Manga(this, provider, manga.series_slug, manga.title.trim()));
```
Comment (ronny1982) [inline web/src/engine/websites/TempleScan.ts:50]: The `trim` is probably redundant as JSON is not expected to contain whitespaces in contrary to HTML.

File: web/src/engine/websites/TempleScan.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
+        const { Season } = await FetchJSON<APIManga>(new Request(new URL(`comic/${manga.Identifier}`, this.apiUrl)));
+        const results = Season.reduce((accumulator: Chapter[], element) => {
+            const chapters = element.Chapter.map(chapter => {
+                const title = chapter.chapter_title ? [chapter.chapter_name, chapter.chapter_title].join(' : ') : chapter.chapter_name;
+                return new Chapter(this, manga, chapter.chapter_slug, title.trim());
+            });
+            accumulator.push(...chapters);
+            return accumulator;
```
Comment (ronny1982) [inline web/src/engine/websites/TempleScan.ts:61]: May use spreading to combine arrays `return [ ...accumulator, ...chapters ];`

**Review by ronny1982 — APPROVED** [review]

File: web/src/engine/websites/TempleScan.ts (hunk 2 above)
Comment (MikeZeDev) [reply web/src/engine/websites/TempleScan.ts:29]: I used it before and i dont know why, on this particular branch, CI was yelling that RegExpSafe was undefined.
I didnt had this error in IDE, and branch was up to date with master.

Comment (MikeZeDev) [reply web/src/engine/websites/TempleScan.ts]: Told you > 'Error: ReferenceError: RegExpSafe is not defined'

**Pull request merged**